### ISSUE TITLE
MethodCallExpr.clone() loses the parent reference

### ISSUE THREAD

**Issue by jlerbsc** [javaparser/javaparser#2292]
Not sure it's a bug but when we clone MethodCallExpr JavaParser loses the parent reference.

    String sample = " { System.out.println(\"\"); }";
    Statement stmt = parseStatement(sample);
    Optional<MethodCallExpr> expr = stmt.findFirst(MethodCallExpr.class);
    assertTrue(expr.isPresent());
    assertTrue(expr.get().getParentNode().isPresent());
    assertTrue(expr.get().clone().getParentNode().isPresent()); // failed

**Comment by matozoid** [comment]
That's correct! JavaParser goes through a lot of trouble to make sure the parent/child relationship stays correct. When you clone a node, they could both have the same parent, but they can't both be a child of the parent. Like: if I clone the expression `1+1` in `return 1+1;` it cannot be stored in the parent. What should it be? `return 1+1;1+1;`?

So the idea is: when you clone, you get a deep copy of the node. It is a stand alone AST. There are a lot of setters and adders in the original AST that you can use to put it back exactly where you want it.

I should write some documentation about this. It's all very logical but somewhat surprising.

**Comment by jlerbsc** [comment]
OK thanks. It's a good explanation ;-)
